cmd/snap/inhibit.go

```diff
@@ -37,15 +37,29 @@ import (
 )
 
 var runinhibitWaitWhileInhibited = runinhibit.WaitWhileInhibited
+var runinhibitIsLocked = runinhibit.IsLocked
 
 // errSnapRefreshConflict indicates that a retry is needed because snap-run
 // might have started without a hint lock file and now there is an ongoing refresh
 // which could alter the current snap revision.
 var errSnapRefreshConflict = fmt.Errorf("snap refresh conflict detected")
 
+// errInhibitedForRemove indicates that snap is inhibited from running because
+// it is being removed.
+var errInhibitedForRemove = fmt.Errorf("snap is being removed")
+
 // maybeWaitWhileInhibited is a wrapper for waitWhileInhibited that skips waiting
-// if refresh-app-awareness flag is disabled.
+// if refresh-app-awareness flag is disabled and early-exits when snap run is
+// inhibited for removal.
 func maybeWaitWhileInhibited(ctx context.Context, cli *client.Client, snapName string, appName string) (info *snap.Info, app *snap.AppInfo, hintFlock *osutil.FileLock, err error) {
+	inhibitedForRemove, err := isInhibitedForRemove(snapName)
```

Comment:
do we have a race like for refresh if we try to remove just after installing and lock file is not there yet? do we need to change/generalize isSnapRefreshConflictDetected perhaps?

Reply:
No, in case of refresh the issue was TOCTOU since we need to hold the lock.

Here (until we add support for app awareness when removing and `--ignore-running`) we don't need to fully sync with snapd and hold the file lock. If/When we support app awareness for removal, then we definitely need to generalize the file locking implementation to support remove inhibition.

Also, Thanks for bringing this up, because when I looked further, there is another race:
- (snap run) check remove inhibition (not locked)
- (snapd) inhibit for remove
- (snapd) kill running apps through transient cgroup
- (snap run) create transient cgroup
- (snap run) run app
This race be resolved by moving the check for remove inhibition after creating the transient cgroup, no lock holding needed. I will add a test and fix for it shortly.

Reply:
It is not as simple as I thought, due to the existing inhibition complexity. I need to think more deeply about the inhibition logic. Thanks for pointing this out!

Reply:
Updated, I had to refactor the waiting logic to extend it for remove hint.